### ISSUE TITLE
Step in and step over does not work reliably

### ISSUE THREAD

**Issue by tgodzik** [scalameta/metals#2233]
**Describe the bug**

It seems often reported by users that step in/out feature is not working currently 100% of time. The debugger is implemented in bloop and is using the microsoft debug server. Not sure how much is a problem inside Bloop and how much is coming from the debugger library.

Reported originally on gitter by @bjornregnell 

**To Reproduce**

Some user reports:
```
Sometimes (but I'm not sure how to reproduce it consistently) the step-in did not work and the stack trace did not show anything if breakpoints are set e.g. in the println in many. I'll see if I can find a way to consistently reproduce during the weekend. Is a link to a video ok in an issue? It is difficult to report GUI-related stuff in pure text without the live action :) ? @tgodzik

I made a video for my students on how to get the debugger with metals going for 2.13.3 and I ran into all sorts of issues such as debugger going crazy in stepping etc. The video is unfortunately in Swedish but I think maybe the failures are still obvious as you can see where I click and what does not happen. Some of the issues are probably not easy to consistently repeat, but maybe this is still valuable to the community as a kind of interactive user experience test. This was run on an x86 box with Ubuntu 18.04 and a rather fast CPU i7-4790K with 8 threads and 4GHz, using VS Code 1.50.1, Scala 2.13.3, sbt 1.4.2, Metals 1.9.6. Here is the link to the video with some strange Metals/bloop behavior (in Swedish): https://www.youtube.com/watch?v=UKLvd_V36bU&t=256s
Let me know if I can contribute with more details of what happened when the debugger step-into etc did not work.
Especially here at time stamp 32:30 the debugging fails when stepping in and out: https://www.youtube.com/watch?v=UKLvd_V36bU&t=1950s

```

based on code sample:

```scala
object Main {
  def many(s: String, n: Int) = {
    println("I'm now in many")
    s * n
  }

  def greet(who: String, n: Int = 42) = 
    many(s"Hello $who! \n", n)

  def main(args: Array[String]): Unit = {
    val mrPresidentElect = "Joe"
    val msg = greet(mrPresidentElect)
    println(msg)
  }
}
```

**Expected behavior**

Step in/out feature works reliably.

**Screenshots**

https://www.youtube.com/watch?v=UKLvd_V36bU&t=1950s

**Installation:**

- Operating system: macOS, but mostly liekly other systems too
- Editor: Visual Studio Code
- Metals version: v0.9.5

**Search terms**

debugging step in out

**Comment by pvid** [comment]
I'll try to look into this over the weekend. Debugging is a part of Metals/Bloop I am at least somewhat familiar with, hopefully I'll be able to come up with something

**Comment by tgodzik** [comment]
Awesome! Thanks a lot for this! Unfortunately, I do not have any idea what's going on.

I would personally try to use the traces generated by Metals for dap client and server to see where exactly is the issue either Bloop/debug server side or Metals.

https://scalameta.org/metals/docs/contributors/getting-started.html#json-rpc-trace

**Comment by pvid** [comment]
Hi, I tried to reproduce the problem and I had some success, but still cannot do it reliably. 

Firstly, lets address jumping into some unrelated method `genericArrayOps` [at 31:50 in the video](https://www.youtube.com/watch?v=UKLvd_V36bU&feature=youtu.be&t=1910). That has something to do with a stale Rambo compiler. If you notice the state of local variables, it does not reflect the addition of the `many` function (added after 30:15) - it has to use a stale compiled version of `Main.scala`. That would imply that doing a "Step Into" on line 6 would step into the `println` function. That is indeed the case (notice the value of `x` after the step). The project uses Scala 2.13.3 and [`println` is on line 422 in `Predef.scala`](https://github.com/scala/scala/blob/v2.12.12/src/library/scala/Predef.scala#L422). The debugger correctly jumps to line 422. However, metals pulls out line 422 of [`Predef.scala` for version 2.12.12](https://github.com/scala/scala/blob/v2.12.12/src/library/scala/Predef.scala#L422), which is the scala version of the Rambo presentation compiler.

I have no idea why that happened. I managed get to the same state once but I did not manage to replicate it. When the file `Main.scala` was created, it was not part of any build target. After "Import build" command has been run, it has started to belong to a build target, but the cached version of its class file produced by the Rambo compiler has not been invalidated.

The apparent unresponsiveness of Step In/Out/Over around 32:30 is caused by the fact that code that is running is not the code on the screen. When you look at the Call Stack (visible on the screen) the thead is paused inside `Thread.exit()`

**Comment by tgodzik** [comment]
Thanks for the great investigation!

> However, metals pulls out line 422 of Predef.scala for version 2.12.12, which is the scala version of the Rambo presentation compiler.

This might have something to do with https://github.com/scalameta/metals/issues/2212

> The apparent unresponsiveness of Step In/Out/Over around 32:30 is caused by the fact that code that is running is not the code on the screen. When you look at the Call Stack (visible on the screen) the thead is paused inside Thread.exit()

If indeed we are seeing old code then it's possible that we are actually at the end of the main function and that's why there is no way to go anymore. 

I think in case of running from code lenses we are no forcing the compile before, which might cause a race condition.

We probably need to check for running compilations here:

https://github.com/scalameta/metals/blob/33e15a45f92d622034659baebb2debcfaf4f3206/metals/src/main/scala/scala/meta/internal/metals/debug/DebugProvider.scala#L88

similar to how it's done here:

https://github.com/scalameta/metals/blob/42b7322352ae56a1668fbedf90274043645e7cc9/metals/src/main/scala/scala/meta/internal/rename/RenameProvider.scala#L60

I think adding this should solve this issue. Are you ok with checking this out?

**Comment by tgodzik** [comment]
Should be fixed in #2450 - we need to wait on compilation to make sure everything is working ok.
